torch/nativert/executor/memory/LayoutManager.cpp

```diff
@@ -157,6 +161,165 @@ void LayoutManager::populate_tensor_values() {
+  //
+  // in our case, this means that leftmost (on the number line) intervals will
+  // come first, and if the start point of two intervals is the same, they will
+  // be sorted by their relative widths (in increasing order)
+  //
+  // e.g., the ordering for the following usage intervals
+  //
+  // |######1######|
+  //        |######2######|
+  //        |######3#####|
+  //
+  // would be [1,3,2]
+
+  std::multiset<std::pair<size_t, size_t>> intervals;
+
+  planner_.with_plan([&](const LayoutPlan& plan) {
+    // prevent recomputation from occuring
+    c10::FastSet<ValueId> checked_values;
+
+    // check that some arbitrary storage (defined by the allocation start and
+    // the size in bytes) lies within the slice allocated for value_id during
+    // planning.
+    //
+    // if the checks pass, add the interval [alloc_start, alloc_start +
+    // alloc_nbytes] to the set of intervals
+    auto check_allocation_bounds =
+        [&](ValueId value_id, size_t alloc_start, size_t alloc_end) -> void {
+      if (!checked_values.emplace(value_id).second /* already checked */) {
+        return;
+      }
+      auto& alloc = plan.allocations[value_to_vector_idx_map_.at(value_id)];
+      TORCH_CHECK_GE(alloc_start, alloc.offset);
+      TORCH_CHECK_LT(alloc_end, alloc.offset + alloc.size);
+      intervals.emplace(alloc_start, alloc_end);
+    };
+
+    // get the inclusive storage interval for some value (i.e.,
+    // [buffer_storage_start_offset, buffer_storage_start_offset +
+    // storage_nbytes]) that represents the sub-slice of the runtime-managed
+    // buffer allocated to this tensor
+    auto try_get_interval =
+        [&](ValueId value_id) -> std::optional<std::pair<size_t, size_t>> {
+      const auto& iv = parent_frame_.getIValue(value_id);
+      if (!iv.isTensor()) {
+        return std::nullopt;
+      }
+
+      const auto& storage_impl = iv.toTensor().storage().unsafeGetStorageImpl();
+      const auto storage_nbytes = storage_impl->nbytes();
+
+      if (const auto start = layout_buffer_.get_offset_from_ptr(
+              storage_impl->data_ptr().get());
+          start.has_value()) {
+        return std::make_pair(*start, *start + storage_nbytes - 1);
+      }
+
+      return std::nullopt;
+    };
+
+    for (auto v : alive_values) {
+      // sanity check lifetimes to ensure this
+      // value ~should~ be alive at this point
+      const auto& lt = alias_analyzer.lifetime(v);
+      TORCH_CHECK_GE(graph_node_idx, lt.start);
+      TORCH_CHECK_LE(graph_node_idx, lt.end);
+
+      const auto interval = try_get_interval(v->id());
+      if (C10_UNLIKELY(!interval.has_value())) {
+        continue;
+      }
+
+      auto& [v_start, v_end] = *interval;
+
+      // it's possible that v is an alias, in which case
+      // we want to try to get the source (i.e., the value)
+      // that actually owns the storage
+      //
+      // NOTE: it's possible the source is ambiguous, hence
```

Comment:
let's chat offline about the source is ambiguous situation. 
non blocking for this PR tho

Reply:
just an idea. 
aliasing relationship should be know after the first forward run.